devices/raspizerow/I2s/main.cpp

```diff
@@ -0,0 +1,301 @@
+    }
+    curl_global_cleanup();
+}
+
+void writeWavHeader(std::ofstream &file, int bitsPerSample, int dataSize)
+{
+    file.write("RIFF", 4);
+    int chunkSize = 36 + dataSize;
+    file.write(reinterpret_cast<const char *>(&chunkSize), 4);
+    file.write("WAVE", 4);
+    file.write("fmt ", 4);
+    int subchunk1Size = 16; // PCM
+    file.write(reinterpret_cast<const char *>(&subchunk1Size), 4);
+    short audioFormat = 1; // PCM
+    file.write(reinterpret_cast<const char *>(&audioFormat), 2);
+    file.write(reinterpret_cast<const char *>(&channels), 2);
+    file.write(reinterpret_cast<const char *>(&sampleRate), 4);
+    int byteRate = sampleRate * channels * bitsPerSample / 8;
+    file.write(reinterpret_cast<const char *>(&byteRate), 4);
+    short blockAlign = channels * bitsPerSample / 8;
+    file.write(reinterpret_cast<const char *>(&blockAlign), 2);
+    file.write(reinterpret_cast<const char *>(&bitsPerSample), 2);
+    file.write("data", 4);
+    file.write(reinterpret_cast<const char *>(&dataSize), 4);
+}
+
+void recordAudio(snd_pcm_t *capture_handle, snd_pcm_uframes_t period_size)
+{
+    std::vector<char> buffer(period_size * bytesPerSample);
+    std::vector<char> accumulatedBuffer;
+
+    while (running)
+    {
+        rc = snd_pcm_readi(capture_handle, buffer.data(), period_size);
+        if (rc == -EPIPE)
+        {
+            // Handle overrun
+            std::cerr << "Overrun occurred" << std::endl;
+            snd_pcm_prepare(capture_handle);
+        }
+        else if (rc < 0)
+        {
+            std::cerr << "Error from read: " << snd_strerror(rc) << std::endl;
+            break; // Exit the loop on error
+        }
+        else if (rc != (int)period_size)
+        {
+            std::cerr << "Short read, read " << rc << " frames" << std::endl;
+        }
+        else
+        {
+            // Append the captured data to the accumulated buffer
+            accumulatedBuffer.insert(accumulatedBuffer.end(), buffer.begin(), buffer.begin() + rc * bytesPerSample * channels);
+
+            if (accumulatedBuffer.size() >= targetBytes)
+            {
+                audioQueue.push(accumulatedBuffer);
+                accumulatedBuffer.clear();
+            }
+        }
+    }
+}
+
+void handleAudioBuffer()
+{
+    while (running)
+    {
+        std::vector<char> dataChunk;
+        // Accumulate enough data to form a complete WAV file
+        while (dataChunk.size() < targetBytes)
+        {
+            std::vector<char> buffer = audioQueue.pop();
+            dataChunk.insert(dataChunk.end(), buffer.begin(), buffer.end());
+        }
+
+        // Process and send the accumulated data
+        if (!dataChunk.empty())
+        {
+            std::ofstream wavFile("output.wav", std::ios::binary);
+            int bitsPerSample = 32;
+            int dataSize = dataChunk.size();
+            writeWavHeader(wavFile, bitsPerSample, dataSize);
+            wavFile.write(dataChunk.data(), dataSize);
```

Comment:
I think we should not save to the fliesystem, as it won't be supported by the microcpntrollers, just keep the file in a var

Reply:
I agree. In the future we will have to take into consideration the amount of memory the device has. For the raspi definitely not an issue but on some of the smaller microcontrollers it might be?